crates/turborepo-lib/src/package_graph/builder.rs

```diff
@@ -54,9 +56,11 @@ pub enum Error {
     Graph(#[from] graph::Error),
     #[error(transparent)]
     Lockfile(#[from] turborepo_lockfiles::Error),
+    #[error(transparent)]
+    Discovery(#[from] turborepo_discovery::Error),
 }
 
-impl<'a> PackageGraphBuilder<'a> {
+impl<'a> PackageGraphBuilder<'a, ()> {
```

Comment:
I think we can avoid the `build_default`/`build` bifurcation if we eagerly providing the default discovery strategy of `LocalPackageDiscovery`. It does an allocation for cloning the repo root, but IMO that's totally fine since this gets called once per run.

Reply:
My worry with this was that we would potentially do package manager detection twice if you got one by default then overwrite it. I think the answer is to just hoist the package manager up out of the strategy and have the caller pass it in as a required argument.

Reply:
Thought we might be able to do a blanket impl on any type that implements TryInto<T> where T implements PackageDiscovery but in retrospect that T is an unconstrained type param so that was also a no-go.

I may leave this as-is despite the API bifurcation.

Reply:
> My worry with this was that we would potentially do package manager detection twice if you got one by default then overwrite it

Ah yeah, I see now that local discovery requires the package manager. 

Maybe the local discovery constructor should take a "package manager factory" i.e. `impl FnOnce() -> Result<PackageManager, Error>`. It would allow for avoiding double detection via `LocalPackageDiscovery::new(repo_root, || Ok(package_manger)`. It also allows for `FallbackPackageDiscovery` to be constructed without accidentally performing a double discovery: `FallbackPackageDiscovery::new(DaemonPackageDiscovery::new(client), LocalPackageDiscovery::new(repo_root, || PackageManger::get_package_manager(repo_root, None))`